File: server/services/node_spec.py
"""Unified NodeSpec assembly.

A NodeSpec is the wire contract between the editor and the backend's
node registry — everything the editor needs to render a node into the
canvas, palette, and parameter panel:

    {
      "type": "...",
      "displayName": "...",
      "icon": "...",
      "group": ["..."],
      "version": 1,
      "inputs": <JSON Schema 7>,    # input parameter schema (Wave 6 Phase 1)
      "outputs": <JSON Schema 7>,   # runtime output shape (Wave 3)
      "credentials": ["..."],       # provider keys, derived from handler
      "uiHints": {...},             # opt-in panel/widget hints
    }

This module is the single point that fuses the three sources of truth:
    services/node_input_schemas.NODE_INPUT_MODELS
    services/node_output_schemas.NODE_OUTPUT_SCHEMAS
    models/node_metadata.NODE_METADATA

Wave 6 Phase 1 — see C:\\Users\\Tgroh\\.claude\\plans\\typed-splashing-crown.md.
"""

import hashlib
import json
from typing import Any, Optional

from models.node_metadata import fallback_metadata, get_node_metadata
from services.node_input_schemas import (
    NODE_INPUT_MODELS,
    get_node_input_schema,
)
from services.node_output_schemas import (
    NODE_OUTPUT_SCHEMAS,
    get_node_output_schema,
)


_spec_cache: dict[str, dict[str, Any]] = {}
_revision_cache: Optional[str] = None


def get_node_spec(node_type: str) -> Optional[dict[str, Any]]:
    """Return the full NodeSpec for a node type, or None if neither an
    input model, an output schema, nor plugin metadata is registered
    (i.e. unknown type). Cached per-process; bust by restarting the
    server.

    Wave 10.G.3: metadata-only plugin nodes (e.g. calculatorTool and
    the other schema-less tool nodes) emit a spec even without input
    schema — their NodeSpec carries componentKind + handles + color,
    which is everything the editor needs to render and connect them.
    """

    if node_type in _spec_cache:
        return _spec_cache[node_type]

    inputs = get_node_input_schema(node_type)
    outputs = get_node_output_schema(node_type)
    meta = get_node_metadata(node_type)
    if inputs is None and outputs is None and meta is None:
        return None
    if meta is None:
        meta = fallback_metadata(node_type)
    spec: dict[str, Any] = {
        "type": node_type,
        "displayName": meta.get("displayName") or node_type,
        "icon": meta.get("icon", ""),
        "group": meta.get("group", []),
        "description": meta.get("description", ""),
        "version": meta.get("version", 1),
    }
    if inputs is not None:
        spec["inputs"] = inputs
    if outputs is not None:
        spec["outputs"] = outputs
    if "subtitle" in meta:
        spec["subtitle"] = meta["subtitle"]
    if meta.get("uiHints"):
        spec["uiHints"] = meta["uiHints"]

    # Wave 10.A — full visual contract. Only emit fields when seeded so the
    # wire format stays compact and unseeded types keep the pre-10 shape.
    for key in ("color", "componentKind", "handles", "credentials", "hideOutputHandle", "hideInputHandle", "visibility"):
        if key in meta:
            spec[key] = meta[key]

    _spec_cache[node_type] = spec
    return spec


def list_node_types_with_spec() -> list[str]:
    """Stable sorted list of every node type the backend knows about —
    either via an input model, an output schema, or a `register_node()`
    metadata entry. The editor uses this on boot to prefetch every spec
    (including visual-only nodes like `simpleMemory` that have metadata
    but no Pydantic schemas)."""

    from models.node_metadata import NODE_METADATA

    return sorted(set(NODE_INPUT_MODELS.keys()) | set(NODE_OUTPUT_SCHEMAS.keys()) | set(NODE_METADATA.keys()))


def node_spec_revision() -> str:
    """Content hash of the full NodeSpec catalogue.

    The frontend persists NodeSpec entries in localStorage with
    ``staleTime: Infinity``; without a revision the persisted cache
    survives backend deploys that change icons, uiHints, or handles
    (e.g. masterSkill gaining ``hideInputSection``). The revision is
    sent alongside ``list_node_specs`` so the editor can detect drift
    on connect and evict stale entries before refetching.

    Computed once per process and cached — restart busts. Hashes the
    fully-assembled spec dict for every known type, sorted, so the
    output is stable regardless of dict-iteration order. First 16 hex
    chars are plenty to avoid collisions for our scale.
    """
    global _revision_cache
    if _revision_cache is not None:
        return _revision_cache
    catalogue = {nt: get_node_spec(nt) for nt in list_node_types_with_spec()}
    payload = json.dumps(catalogue, sort_keys=True, default=str).encode("utf-8")
    _revision_cache = hashlib.sha256(payload).hexdigest()[:16]
    return _revision_cache


def list_node_groups() -> dict[str, dict[str, Any]]:
    """Per-group index for the component palette.

    Wave 10.B shape::

        {
          "agent":   {"types": [...], "label": "AI Agents",
                      "icon": "🤖", "color": "#bd93f9",
                      "visibility": "normal"},
          "model":   {"types": [...], ...},
          ...
        }

    `types` is the sorted list of node types belonging to the group
    (replaces the 34 hand-rolled `*_NODE_TYPES` arrays scattered across
    the frontend). Remaining keys come from
    ``models.node_metadata.GROUP_METADATA`` and make the palette's
    section headers fully backend-driven — icon, label, color, and
    "normal"-vs-"dev" visibility are all declared in
    ``server/nodes/groups.py``.

    When a group has nodes but no metadata seeded, the entry falls back
    to the raw key as label with no icon/color; the frontend can still
    render it safely.
    """
    from models.node_metadata import get_group_metadata

    types_by_group: dict[str, set[str]] = {}
    for node_type in list_node_types_with_spec():
        spec = get_node_spec(node_type)
        if not spec:
            continue
        for group in spec.get("group", []):
            types_by_group.setdefault(group, set()).add(node_type)

    result: dict[str, dict[str, Any]] = {}
    for group, type_set in sorted(types_by_group.items()):
        meta = get_group_metadata(group) or {}
        result[group] = {
            "types": sorted(type_set),
            "label": meta.get("label") or group,
            "icon": meta.get("icon", ""),
            "color": meta.get("color", ""),
            "visibility": meta.get("visibility", "dev"),
        }
    return result

File: client/src/lib/nodeSpec.ts
/**
 * NodeSpec fetch + resolve utilities.
 *
 * Wave 6 Phase 2. Thin wrapper around TanStack Query's `fetchQuery`
 * mirroring the Wave 3 pattern at InputSection.tsx:28-36 — no new hook
 * file, just a shared async helper so every consumer can opt in behind
 * the VITE_NODESPEC_BACKEND flag without duplicating boilerplate.
 *
 * Query key convention: ['nodeSpec', nodeType]
 * Stale time: Infinity (node shapes only change with a deploy).
 *
 * See C:\\Users\\Tgroh\\.claude\\plans\\typed-splashing-crown.md.
 */

import { useEffect, useSyncExternalStore } from 'react';
import { hashKey, useQuery, type UseQueryResult } from '@tanstack/react-query';
import { nodeSpecToDescription, type NodeSpec } from '../adapters/nodeSpecToDescription';
import type { INodeTypeDescription } from '../types/INodeProperties';
import { featureFlags } from './featureFlags';
import { BRAND_STORAGE_KEYS, readAndMigrateStorageValue } from './brandStorage';
import { queryClient } from './queryClient';
import { GC_TIME, STALE_TIME } from './queryConfig';
import { useWebSocket } from '../contexts/WebSocketContext';

type SendRequest = (type: string, data?: any) => Promise<any>;

export const nodeSpecQueryKey = (nodeType: string) => ['nodeSpec', nodeType] as const;

/**
 * Fetch a single NodeSpec via the WS handler, caching per node type.
 * Returns `null` when the type is unknown on the backend — callers fall
 * back to the legacy `nodeDefinitions/*` entry.
 */
export async function fetchNodeSpec(
  nodeType: string,
  sendRequest: SendRequest,
): Promise<NodeSpec | null> {
  return queryClient.fetchQuery({
    queryKey: nodeSpecQueryKey(nodeType),
    queryFn: async () => {
      try {
        const response = await sendRequest('get_node_spec', { node_type: nodeType });
        return (response?.spec ?? null) as NodeSpec | null;
      } catch {
        return null;
      }
    },
    staleTime: STALE_TIME.FOREVER,
    // useNodeSpec subscribes via useSyncExternalStore (no observer), so
    // without an explicit gcTime override every spec entry is evicted
    // ~5 min after prefetch lands. That makes node icons + handles
    // disappear on every idle canvas. Keep the entries forever; the
    // persistor in lib/queryPersist.ts also writes them to localStorage.
    gcTime: GC_TIME.FOREVER,
  });
}

const NODESPEC_REVISION_STORAGE_KEYS = BRAND_STORAGE_KEYS.nodeSpecRevision;

/**
 * Prefetch every node type the backend knows about, in a single idle
 * burst at editor boot. Non-blocking — failures are swallowed so a flaky
 * WS doesn't block UI. Matches Activepieces' summary+lazy pattern.
 *
 * Self-busts the persisted spec cache when the backend catalogue
 * changes. The backend returns a content-hash `revision` derived from
 * the full NodeSpec catalogue (icons, uiHints, handles, schemas);
 * without this, persisted entries with `staleTime: FOREVER` survive
 * server deploys and the editor keeps rendering pre-deploy shapes
 * (e.g. masterSkill missing its `hideInputSection` hint). On revision
 * mismatch we evict the relevant prefixes before refetching so the
 * persistor overwrites stale localStorage entries on the next sync.
 */
export async function prefetchAllNodeSpecs(sendRequest: SendRequest): Promise<void> {
  try {
    const response = await sendRequest('list_node_specs', {});
    const nodeTypes: string[] = response?.node_types ?? [];
    const revision: string | undefined = response?.revision;

    if (revision && typeof window !== 'undefined') {
      const lastSeen = readAndMigrateStorageValue(
        window.localStorage,
        NODESPEC_REVISION_STORAGE_KEYS,
      );
      if (lastSeen !== revision) {
        queryClient.removeQueries({ queryKey: ['nodeSpec'] });
        queryClient.removeQueries({ queryKey: ['nodeGroups'] });
        window.localStorage.setItem(NODESPEC_REVISION_STORAGE_KEYS.canonical, revision);
      }
    }

    await Promise.all(nodeTypes.map(t => fetchNodeSpec(t, sendRequest)));
  } catch {
    // Prefetch is best-effort — logged by the WS layer, not us.
  }
}

/**
 * Synchronous read of a previously-cached NodeSpec. Used by the
 * non-reactive merge helpers (`resolveNodeDescription`,
 * `isNodeInBackendGroup`). Prefer `useNodeSpec` in React components
 * so the render updates when prefetch lands.
 */
export function getCachedNodeSpec(nodeType: string): NodeSpec | null {
  return queryClient.getQueryData<NodeSpec | null>(nodeSpecQueryKey(nodeType)) ?? null;
}

/**
 * Reactive spec subscription via the QueryCache as an external store.
 *
 * Previous implementation opened a `useQuery` observer per call site;
 * with ~80 palette items + N canvas nodes that produced 80+N observers,
 * each woken up on every cache write. The slice-subscription pattern
 * here re-renders only consumers of the *specific* spec key that
 * changed — observer count drops to zero. Pattern documented at
 * https://react.dev/reference/react/useSyncExternalStore and called
 * out by TanStack's docs as the canonical escape hatch from useQuery.
 *
 * Cache population is owned by `prefetchAllNodeSpecs` (boot once) and
 * the persisted localStorage hydration set up in lib/queryPersist.ts;
 * a missing key triggers a one-shot lazy fetch via useEffect rather
 * than a long-lived observer.
 */
export function useNodeSpec(nodeType: string | undefined | null): NodeSpec | null {
  const { sendRequest, isReady } = useWebSocket();
  const key = nodeSpecQueryKey(nodeType ?? '__none__');

  const targetHash = hashKey(key);
  const subscribe = (onChange: () => void) => {
    const unsub = queryClient.getQueryCache().subscribe((event) => {
      if (event.query.queryHash === targetHash) onChange();
    });
    return unsub;
  };

  const getSnapshot = () =>
    queryClient.getQueryData<NodeSpec | null>(key) ?? null;

  const data = useSyncExternalStore(subscribe, getSnapshot, getSnapshot);

  useEffect(() => {
    if (!nodeType || !isReady) return;
    if (queryClient.getQueryData(key) !== undefined) return;
    void fetchNodeSpec(nodeType, sendRequest);
    // eslint-disable-next-line react-hooks/exhaustive-deps -- key derives from nodeType; including it is redundant.
  }, [nodeType, isReady, sendRequest]);

  return data;
}

/**
 * Reactive node-groups subscription. Same WS-in-queryFn pattern as
 * `useNodeSpec`. Returns the full `UseQueryResult` so callers can
 * branch on `data` / `isPending` / `error` and render a proper loading
 * state rather than masking missing data. Matches TkDodo's recommended
 * status-check pattern: check data first, error second, loading last.
 */
export function useNodeGroups(): UseQueryResult<Record<string, NodeGroupEntry>> {
  const { sendRequest, isReady } = useWebSocket();
  return useQuery<Record<string, NodeGroupEntry>>({
    queryKey: nodeGroupsQueryKey,
    queryFn: async () => {
      const response = await sendRequest('get_node_groups', {});
      return (response?.groups ?? {}) as Record<string, NodeGroupEntry>;
    },
    enabled: isReady,
    staleTime: STALE_TIME.FOREVER,
    gcTime: GC_TIME.FOREVER,
  });
}

/** Wire shape for one entry in the GET /api/schemas/nodes/groups
 *  response (Wave 10.B): per-group palette metadata + member types. */
export interface NodeGroupEntry {
  types: string[];
  label: string;
  icon: string;
  color: string;
  visibility: 'all' | 'normal' | 'dev';
}

export const nodeGroupsQueryKey = ['nodeGroups'] as const;

/**
 * Wave 10.B: fetch the full per-group palette index from the backend.
 * Cached forever (group metadata only changes with a redeploy). The
 * frontend ComponentPalette consumes this directly — no hand-rolled
 * `CATEGORY_ICONS` / `labelMap` / `SIMPLE_MODE_CATEGORIES` tables.
 */
export async function fetchNodeGroups(
  sendRequest: SendRequest,
): Promise<Record<string, NodeGroupEntry>> {
  return queryClient.fetchQuery({
    queryKey: nodeGroupsQueryKey,
    queryFn: async () => {
      const response = await sendRequest('get_node_groups', {});
      return (response?.groups ?? {}) as Record<string, NodeGroupEntry>;
    },
    staleTime: STALE_TIME.FOREVER,
    gcTime: GC_TIME.FOREVER,
  });
}

/**
 * Wave 10.E: enumerate every NodeSpec the cache currently holds.
 * Used by component palette / drag-drop / execution routing to derive
 * filtered lists from spec metadata instead of importing per-family
 * type arrays from `nodeDefinitions/*`.
 */
export function listCachedNodeSpecs(): NodeSpec[] {
  const all = queryClient.getQueriesData<NodeSpec | null>({ queryKey: ['nodeSpec'] });
  return all
    .map(([, spec]) => spec)
    .filter((s): s is NodeSpec => !!s);
}

/**
 * Wave 10.E: types whose backend spec lists ``group`` includes the
 * given group key. Returns an empty array until prefetch lands; callers
 * that need a synchronous answer should rely on the spec-driven
 * dispatcher rather than this enumeration.
 */
export function getNodeTypesInGroup(group: string): string[] {
  return listCachedNodeSpecs()
    .filter(s => (s.group ?? []).includes(group))
    .map(s => s.type);
}

/**
 * Wave 10.E: types whose backend spec carries the given componentKind.
 */
export function getNodeTypesWithKind(kind: NonNullable<NodeSpec['componentKind']>): string[] {
  return listCachedNodeSpecs()
    .filter(s => s.componentKind === kind)
    .map(s => s.type);
}

/**
 * Wave 6 Phase 5.b: backend-group membership check.
 *
 * Returns `undefined` when the NodeSpec isn't cached yet (caller falls
 * back to its legacy `*_NODE_TYPES` array), `true`/`false` when the
 * cached spec's `group` array decides. Lets components retire local
 * helper arrays without introducing a hard dependency on prefetch
 * ordering — when the flag is off and prefetch hasn't run, the legacy
 * path runs unchanged.
 */
export function isNodeInBackendGroup(
  nodeType: string | null | undefined,
  group: string,
): boolean | undefined {
  if (!nodeType) return false;
  const spec = getCachedNodeSpec(nodeType);
  if (!spec) return undefined;
  return (spec.group ?? []).includes(group);
}

/**
 * Wave 7 flag-gated resolver.
 *
 * Flag OFF: returns `localFallback` unchanged.
 * Flag ON + spec cold: falls back to `localFallback` so the editor
 *   never renders an empty panel during prefetch warmup.
 * Flag ON + spec warm: returns the backend NodeSpec adapted to the
 *   INodeTypeDescription shape. Local UX-only hints that are absent
 *   from the NodeSpec (starter code defaults, placeholder JSON) are
 *   merged in per-property so the adapter never regresses UX.
 *
 * Merge rules (backend wins on schema, local wins on UX):
 *   - Schema fields (type, options, displayOptions, typeOptions,
 *     validation, required, description): always backend when present
 *   - UX fields (placeholder): backend when non-empty, else local
 *   - default: backend when non-empty, else local (preserves starter
 *     code + example JSON blobs that live only in the frontend)
 */
export function resolveNodeDescription(
  nodeType: string,
  localFallback?: INodeTypeDescription | null,
): INodeTypeDescription | null {
  if (!featureFlags.nodeSpecBackend) {
    return localFallback ?? null;
  }
  const spec = getCachedNodeSpec(nodeType);
  if (!spec) return localFallback ?? null;

  const backend = nodeSpecToDescription(spec);
  if (!localFallback) return backend;

  // Per-property merge: keep backend as authoritative but pull UX
  // niceties (placeholder + non-empty default) from local when the
  // backend version is empty. Unknown local properties (no matching
  // backend entry) are dropped - backend is the schema SSOT.
  const localByName = new Map(
    (localFallback.properties ?? []).map((p) => [p.name, p]),
  );
  const mergedProperties = (backend.properties ?? []).map((bp) => {
    const lp = localByName.get(bp.name);
    if (!lp) return bp;
    const merged = { ...bp };
    const bpDefault = (bp as any).default;
    const bpDefaultEmpty =
      bpDefault === undefined ||
      bpDefault === null ||
      bpDefault === '' ||
      (typeof bpDefault === 'object' &&
        !Array.isArray(bpDefault) &&
        Object.keys(bpDefault).length === 0);
    if (bpDefaultEmpty && (lp as any).default !== undefined) {
      (merged as any).default = (lp as any).default;
    }
    if (!bp.placeholder && lp.placeholder) merged.placeholder = lp.placeholder;
    if (!bp.description && lp.description) merged.description = lp.description;
    return merged;
  });

  // Wave 10.B: backend NodeSpec is the sole source for top-level
  // visual metadata (icon, subtitle, description, color). Local
  // nodeDefinitions/*.ts entries carry no icons anymore, so we just
  // pass the backend values through and only preserve local
  // `defaults.color` when the backend doesn't declare one — color is
  // the last remaining UX field that a few specialised agent configs
  // still set locally.
  return {
    ...backend,
    defaults: {
      ...localFallback.defaults,
      ...backend.defaults,
      color: backend.defaults?.color || localFallback.defaults?.color,
    },
    properties: mergedProperties,
  };
}
